vnext/Shared/Threading/BatchingQueueThread.cpp

```diff
@@ -57,20 +56,56 @@ void BatchingQueueThread::PostBatch() noexcept {
   }
 }
 
-void BatchingQueueThread::onBatchComplete() noexcept {
+void BatchingQueueCallInvoker::onBatchComplete() noexcept {
   std::scoped_lock lck(m_mutex);
   PostBatch();
 }
 
+void BatchingQueueCallInvoker::quitSynchronous() noexcept {
+  std::scoped_lock lck(m_mutex);
+  PostBatch();
+  m_queueThread->quitSynchronous();
+}
+
+void BatchingQueueCallInvoker::invokeSync(std::function<void()> &&func) noexcept {
+  assert(false && "Not supported");
+  std::terminate();
+}
+
+BatchingQueueThread::BatchingQueueThread(
+    std::shared_ptr<facebook::react::MessageQueueThread> const &queueThread) noexcept {
+  m_batchingQueueCallInvoker = std::make_shared<BatchingQueueCallInvoker>(queueThread);
+  m_callInvoker = m_batchingQueueCallInvoker;
+}
+
+// Called from the JS thread, once the instance has had a chance to setup the NativeToJsBridge and calls to
+// getDecoratedNativeCallInvoker will work.
+void BatchingQueueThread::decoratedNativeCallInvokerReady(
+    std::weak_ptr<facebook::react::Instance> wkInstance) noexcept {
+  m_callInvoker->invokeAsync([wkInstance, this] {
+    if (auto instance = wkInstance.lock()) {
+      m_callInvoker = instance->getDecoratedNativeCallInvoker(m_callInvoker);
```

Comment:
>m_callInvoker [](start = 6, length = 13)

Should we protect access to this variable? The runOnQueue below can be called froma different thread.

Reply:
It should be safe.  But the logic to verify that is pretty convoluted -- and so not safe for future changes.  I'll just capture a ref to it.

Reply:
This should be ok.  The only place that this variable changes is from this code, and the actual value change will be on the UI thread. All other reads of the variable are on the UI thread.  The read in this code that is on the JS thread will always be before safe since the task to modify it is kicked off after that access.

Reply:
No, it is not OK. Previously this code was called under the lock that has a guarantee of the thread safety. After this change we are going to have a race condition. Please add the lock back.